### ISSUE TITLE
Configure Network connection to be an Internet interface

### ISSUE THREAD

**Issue by steglicd** [freedombox/Plinth#157]
A configuration option should be added to mark a network connection as the "upstream" or "internet" interface. Only one interface can be the Internet Interface. If a Interface is marked as "internet", connection should be configured as "shared" (from networkmanager point of view).

**Comment by SunilMohanAdapa** [comment]
Due to recent issue with freedombox-setup before the 0.5 release, I ended up investigating the shared mode of the network-manager.  Here is how it works according to a comment in code.

> This connection specifies configuration that allows other computers to
> connect through it to the default network (usually the Internet).  The
> connection's interface will be assigned a private address, and a DHCP server,
> caching DNS server, and Network Address Translation (NAT) functionality will
> be started on this connection's interface to allow other devices to connect
> through that interface to the default network.

So, to implement the solution, we need to:

1. We should provide a 'Shared' mode, apart from 'Automatic' and 'Manual' methods of providing IPv4 addressing.  In this mode, all the other fields should be disabled as network manager automatically assign the addresses etc.  This is how GNOME's network manager does it.

2. When the user selects 'shared' mode for a connect, we need to set 'ipv4.method = shared' on the that connection.  The user will set this on 'internal' interfaces and not on Internet connected interface. Network manager will take care of setting up DNS, DHCP and NAT forwarding for this interface.

@steglicd , let me know if you are interested in picking this up.

**Comment by steglicd** [comment]
@SunilMohanAdapa yes, after I finished PPPoE integration I would like to work on this.

I think from the UI point of view the user should decide: 
* is it a interface connected TO the Internet
* or is it a interface where I want to connect my LAN clients to

If it is a interface where the LAN clients are connected to, we may offer DHCP range configuration. Otherwise nothing else should be required for configuration.

**Comment by SunilMohanAdapa** [comment]
Wait! This functionality seems to be already available in FreedomBox with the original patch submitted by @jvalleroy .  Although I reviewed the patch, I totally forgot about this.  @steglicd could you please confirm this is working?

**Comment by steglicd** [comment]
@SunilMohanAdapa 
I will test this. I have seen this option already but I did not get the usecase of the "shared" feature. Because I do not "share" this connection but share other connections via this connection. Maybe we could think about naming and usability? Because I think the user only knows two things:
* this is the wire I connect to the Internet (upstream)
* this is the wire I connect to my LAN clients (downstream)

I'm not sure if the user will understand networkmanagers logic about "shared" mode?

**Comment by SunilMohanAdapa** [comment]
On 08/18/2015 02:49 PM, Daniel Steglich wrote:
> @SunilMohanAdapa <https://github.com/SunilMohanAdapa>
> I will test this. I have seen this option already but I did not get the
> usecase of the "shared" feature. Because I do not "share" this
> connection but share other connections via this connection. Maybe we
> could think about naming and usability? Because I think the user only
> knows two things:
> 
>   * this is the wire I connect to the Internet (upstream)
>   * this is the wire I connect to my LAN clients (downstream)
> 
> I'm not sure if the user will understand networkmanagers logic about
> "shared" mode?

I agree.

How about we make the IPv4 connection method a set of radio buttons and
write detailed description on what each of them are?  We can explain for
the shared mode how it will work if selected.

-- 
Sunil
